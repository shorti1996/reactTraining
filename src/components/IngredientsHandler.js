import React from 'react';

const INGREDIENTS = [
    { label: 'Tomato', type: 'tomato' },
    { label: 'Bacon', type: 'bacon' },
    { label: 'Chesse', type: 'cheese' },
    { label: 'Salad', type: 'salad' },
    { label: 'Meat', type: 'meat' },
];

const IngredientsHandler = ({ ingredientAdd, ingredientRemove })  => {
    return (
        <ul className="ingredient-handler">
            {
                INGREDIENTS.map((ingredient, index) => (
                    <li key={index}>
                        {ingredient.label}
                        <button onClick={() => ingredientAdd(ingredient.type)}>+</button>
                        <button onClick={() => ingredientRemove(ingredient.type)}>-</button>
                    </li>
                ))
            }
        </ul>
    );
};

export default IngredientsHandler;