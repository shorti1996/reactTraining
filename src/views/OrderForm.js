import React, { Component } from 'react';
import SubmitOrder from '../services/SubmitOrder';

class OrderForm extends Component {
    state = {
        name: '',
        address: '',
        phone: '',
        email: '',
    };

    handleSubmit = event => {
        event.preventDefault();
        // console.log(this.state);
        // console.log(this.props.location.state);
        SubmitOrder(this.state.name, this.state.surname, this.props.location.state.ingredients)
    };

    handleInputChange = event => {
        const target = event.target;
        const name = target.name;

        this.setState({
            [name]: target.value
        });
    };
    
    render() {
        return (
            <form className="order-form" onSubmit={this.handleSubmit} >
                <div>
                    <label>Name</label>
                    <input name="name" value={this.state.name} onChange={this.handleInputChange} />
                </div>
                <div>
                    <label>Address</label>
                    <input name="address" value={this.state.address} onChange={this.handleInputChange} />
                </div>
                <div>
                    <label>Phone</label>
                    <input name="phone" value={this.state.phone} onChange={this.handleInputChange} />
                </div>
                <div>
                    <label>Email</label>
                    <input name="email" value={this.state.email} onChange={this.handleInputChange} />
                </div>
                <button type="submit">Submit</button>
            </form>
        );
    };
};

export default OrderForm;