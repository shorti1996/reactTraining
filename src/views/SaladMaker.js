import React, { Component } from 'react';
import Salad from '../components/Salad';
import prepareIngredients from '../utils/prepareIngredients';
import IngredientsHandler from '../components/IngredientsHandler';
import OrderForm from '../components/OrderForm';

class SaladMaker extends Component {
    state = {
        ingredients: {
            tomato: 1,
            bacon: 2,
            cheese: 3,
            salad: 4,
            meat: 5,
        },
        isReadyToOrder: false,
    };

    ingredientAdd = type => {
        this.setState({
            ingredients: {
                ...this.state.ingredients,
                [type]: this.state.ingredients[type] + 1
            }
        })
    };

    ingredientRemove = type => {
        if (this.state.ingredients[type] <= 0) return;
        this.setState({
            ingredients: {
                ...this.state.ingredients,
                [type]: this.state.ingredients[type] - 1
            }
        })
    };

    orderSalad = () => { this.setState({ isReadyToOrder: true }) };

    render() {
        const preparedIngredients = prepareIngredients(this.state.ingredients);

        return (
            <div className="salad-maker">
                <h1>SaladMaker</h1>
                <Salad ingredients={preparedIngredients} />
                <IngredientsHandler ingredientAdd={this.ingredientAdd} ingredientRemove={this.ingredientRemove} />
                { this.state.isReadyToOrder
                    ? <OrderForm /> 
                    : <button onClick={this.orderSalad}>Next step</button> 
                }
            </div>
        );
    };
};

export default SaladMaker;