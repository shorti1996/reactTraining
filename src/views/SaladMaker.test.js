import React from 'react';
import SaladMaker from './SaladMaker'
import { shallow } from 'enzyme';
import IngredientsHandler from '../components/IngredientsHandler';
import OrderForm from '../components/OrderForm';

describe('SaladMaker', () => {
    it('should render without crash', () => {
        const sut = createSut();
        
        expect(sut.isEmptyRender()).toBe(false);
    });

    it('should match snapshot', () => {
        const sut = createSut();

        expect(sut).toMatchSnapshot();
    });

    let sut;

    beforeEach(() => {
        sut = createSut();
    })

    describe('salad ingredients', () => {

        it('should pass empty ingredients to salad by default', () => {
            expect(sut.find('Salad').props().ingredients).toEqual([]);
        })

        it('should pass incremented count of specified ingredient when such ingredient not exists', () => {
            const INGREDIENT = 'INGREDIENT';
            sut.setState({ingredients: {}});

            sut.find(IngredientsHandler).props().ingredientAdd(INGREDIENT);

            expect(sut.find('Salad').props().ingredients).toEqual([INGREDIENT]);
        })

        it('should pass incremented count of specified ingredient when such ingredient exists', () => {
            const INGREDIENT = 'INGREDIENT';
            sut.setState({ingredients: { [INGREDIENT]: 1 }});
            
            sut.find(IngredientsHandler).props().ingredientAdd(INGREDIENT);

            expect(sut.find('Salad').props().ingredients).toEqual([INGREDIENT, INGREDIENT]);
        })
    });

    describe('order form', () => {
        it('should form be hidden by default', () => {
            expect(sut.find('OrderForm').exists()).toBe(false);
        });

        it('should be visible after click on order button', () => {
            const button = sut.find('button');
            button.simulate('click');
            expect(sut.find('OrderForm').exists()).toBe(true);
        });
    });
    
    const createSut = () => shallow(<SaladMaker />);
});

// describe('OrderForm', () => {
//     let sut;

//     beforeEach(() => {
//         sut = createSut();
//     })

//     it('should form be hidden by default', () => {
//         expect(sut.find('Salad').props().ingredients).toEqual([]);
//     });

//     const createSut = () => shallow(<OrderForm />);
// });