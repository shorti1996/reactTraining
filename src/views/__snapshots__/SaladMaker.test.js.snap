// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaladMaker should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SaladMaker />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          SaladMaker
        </h1>,
        <IngredientsHandler
          ingredientAdd={[Function]}
          ingredientRemove={[Function]}
        />,
        <button
          onClick={[Function]}
        >
          Next step
        </button>,
        <Salad
          ingredients={Array []}
        />,
      ],
      "className": "salad-maker",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "SaladMaker",
        },
        "ref": null,
        "rendered": "SaladMaker",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "ingredientAdd": [Function],
          "ingredientRemove": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Next step",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Next step",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "ingredients": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            SaladMaker
          </h1>,
          <IngredientsHandler
            ingredientAdd={[Function]}
            ingredientRemove={[Function]}
          />,
          <button
            onClick={[Function]}
          >
            Next step
          </button>,
          <Salad
            ingredients={Array []}
          />,
        ],
        "className": "salad-maker",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SaladMaker",
          },
          "ref": null,
          "rendered": "SaladMaker",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "ingredientAdd": [Function],
            "ingredientRemove": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Next step",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Next step",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "ingredients": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
